### DISCUSSION TITLE
TuYa TS0601_din_2 ("energy meter") "bidirectional" power/energy issue

### DISCUSSION THREAD

**Discussion by fred-c1** [Koenkk/zigbee2mqtt#16438] [Devices]
*********** Intro:
"bidirectional" (foward and backward) power/energy measurement is very usefull when you are connected to the electricity grid and are also producing some energy (e.g. solar panels + inverter)
For forward power, I am expecting a posive value of "power" (e.g. "82" for 82 watt)
For reverse power, I am expecting a negative value of "power" (e.g. "-82" for -82 watt)
The information provide on the LCD screen of the meter is "correct":
- For forward power: just the power value
- For reverse power: power value + "right to left arrow" to indicated reverse/negative

*********** Details:
I received the "energy meter" a few days ago:
https://www.aliexpress.com/item/1005004997500200.html
description = "Tuya Smart ZigBee Energy Meter Bidirectional Single Phase 80A Din Rail Power Monitor Wattmeter Voltmeter Ampermeter 110V 240V"
The important word here is "bidirectional"
store = MatSeePlus Official Store

I am using an external "zigbee-herdsman" converter  based on this one:
https://www.zigbee2mqtt.io/devices/TS0601_din_2.html
The only change I made is the "zigbee ID".
The "ID" of my energy meter is '_TZE200_rhblgy0z'

*********** test using forward power / energy
LCD screen on the "energy meter" = just the value (81 Watt)
zigbee data =
{"current":0.35,"energy":0,"last_seen":"2023-01-23T19:56:51+01:00","linkquality":26,"meter_id":"22492596","power":81,"state":"ON","threshold_1":"over_current_threshold","threshold_1_protection":"ON","threshold_1_value":80,"threshold_2_value":0,"voltage":240}

*********** test using reverse power / energy
LCD screen on the "energy meter" =  value (81 Watt) + arrow from right to left (indicating reverse power/energy)
zigbee data =
{"current":0.35,"energy":0,"last_seen":"2023-01-23T19:53:23+01:00","linkquality":49,"meter_id":"22492596","power":81,"state":"ON","threshold_1":"over_current_threshold","threshold_1_protection":"ON","threshold_1_value":80,"threshold_2_value":0,"voltage":240.6}

*********** Conclusion:
This energy meter is not bidirectional since the zigbee data is the same for forward and backward power/energy

*********** Questions:
Could this be a "zigbee2MQTT converter" issue ?
For example if the converter "removes" the sign of the value????  ("-81" becomes "81").
I don't have a Tuya gateway. I am only testing with "zigbee2MQTT.
I am also asking the vendor to help.

**Comment by mztulip** [comment]
Any progress with this? I want to buy this energy meter and I want to know if is it possible to correct.

**Comment by fred-c1** [comment]
@buleks    long story but in short the new version of the meter should have "bidirectional real time power".
However, you should confirm this with the vendor
and the measurements update rate is only "about every 2 minutes".

About "bidirectional real time power":
This meter 
https://www.aliexpress.com/item/1005005466228073.html
has just been added to Z2M dev branch.
https://github.com/Koenkk/zigbee2mqtt/issues/18130
It has "bidirectional real time power" and the update rate is "about every 3 seconds" when the current is above "0.1A".
The update rate is "about every 37 seconds" when the current is below "0.1A".

**Reply by Gambix91** [reply]
So you have installed this new smart meter, and it shows you in Home assistant positive number when you pull energy from the grid and negative numbers when your solar is producing more energy than your house is requiring, so the difference go back in the grid?

**Comment by fred-c1** [comment]
Correct
